qiskit/circuit/library/data_preparation/state_preparation.py

```diff
@@ -414,3 +439,137 @@ def _multiplex(self, target_gate, list_of_angles, last_cnot=True):
             circuit.append(CXGate(), [msb, lsb])
 
         return circuit
+
+
+class Generalized_Uniform_Superposition_Gate(Gate):
+    """
+    Class that implements a generalized uniform superposition state, using n qubits, following the Shukla-Vedula algorithm [SV24].
+    """
+
+    def __init__(
+        self,
+        params: Union[str, list, int, Statevector],
+        M : int = 2,
+        num_qubits: Optional[int] = None,
```

Comment:
I am not understanding why params is here. I know they are passed to the parent but this is not something that is used here right - your only params are M. In your test, where you are doing `Generalized_Uniform_Superposition_Gate(M, n)` the M there goes to params (the first parameter in the list heer), n ends up in M (the 2nd position) and num_qubits is left as None.

Hence this call fails in the parent `super().__init__("USup", num_qubits, [])` as it passes None (as far as I can see).

My take would be to remove params from here, have M first here, not have a default value so you have to provide a value, and then num_qubits as you have it.

Next move this call `super().__init__("USup", num_qubits, [])` to the end of the method and change it to ``super().__init__("USup", self._num_qubits, [M])` This way you always pass a valid value of num qubits to the parent since you compute it if its None based on M. The as M is a parameter make that a single value in the list passed as parameters - this is kinda the equivalent of what params is to the StatePreparation. You can see how its handled there.

As to ddt you are not using it here. You can parameterize the test and even create the data so you do not need a for loop rather ddt will create a single test for each data entry there. That is better as all the tests will always run - in the for loop, as you have it, it will stop at the first failure. Usually we like to run all tests and if a test had independent parts like this, but is in one test, it would be done with the self.subTest. This is something that can be sorted once the code is running properly - I think what I am suggesting above will fix things.